src/streaming/s3.connection.cpp

```diff
@@ -0,0 +1,254 @@
+#include "macros.hh"
+#include "s3.connection.hh"
+
+#include <miniocpp/utils.h>
+
+#include <list>
+#include <sstream>
+#include <string_view>
+
+zarr::S3Connection::S3Connection(const std::string& endpoint,
+                                 const std::string& access_key_id,
+                                 const std::string& secret_access_key)
+{
+    minio::s3::BaseUrl url(endpoint);
+    url.https = endpoint.starts_with("https");
+
+    provider_ = std::make_unique<minio::creds::StaticProvider>(
+      access_key_id, secret_access_key);
+    client_ = std::make_unique<minio::s3::Client>(url, provider_.get());
+
+    CHECK(client_);
+}
+
+bool
+zarr::S3Connection::check_connection()
+{
+    return static_cast<bool>(client_->ListBuckets());
+}
+
+bool
+zarr::S3Connection::bucket_exists(std::string_view bucket_name)
+{
+    EXPECT(!bucket_name.empty(), "Bucket name must not be empty.");
+
+    minio::s3::BucketExistsArgs args;
+    args.bucket = bucket_name;
+
+    auto response = client_->BucketExists(args);
+    return response.exist;
+}
+
+bool
+zarr::S3Connection::object_exists(std::string_view bucket_name,
+                                  std::string_view object_name)
+{
+    EXPECT(!bucket_name.empty(), "Bucket name must not be empty.");
+    EXPECT(!object_name.empty(), "Object name must not be empty.");
+
+    minio::s3::StatObjectArgs args;
+    args.bucket = bucket_name;
+    args.object = object_name;
+
+    auto response = client_->StatObject(args);
+    // casts to true if response code in 200 range and error message is empty
+    return static_cast<bool>(response);
+}
+
+std::string
+zarr::S3Connection::put_object(std::string_view bucket_name,
+                               std::string_view object_name,
+                               std::span<uint8_t> data)
+{
+    EXPECT(!bucket_name.empty(), "Bucket name must not be empty.");
+    EXPECT(!object_name.empty(), "Object name must not be empty.");
+    EXPECT(!data.empty(), "Data must not be empty.");
+
+    minio::utils::CharBuffer buffer((char*)const_cast<uint8_t*>(data.data()),
```

Comment:
`const_cast` is considered bad because it can break the safety guarantees provided by const in C++. The const qualifier ensures that a value cannot be modified after it is initialized, which helps avoid accidental changes and makes code more predictable and easier to reason about.

The only reasonable explanation for using it is interfacing with legacy code that doesn't respect const correctness or maybe working with APIs that incorrectly omit const qualifiers in places where they should, but both cases are uncommon.

Why did you use it in the first place?